BFD-2737: Adjust github hook installation and move/rename githook secrets file

lsmitchell opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BFD-2737](https://jira.cms.gov/browse/BFD-2737)

**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
As a BFD contributor, I want the pre-commit and pre-push hooks to be sane so that I can use them with sanity.

---

### What Does This PR Do?
<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->

- Moves the git hooks into their own files instead of shoved into the parent pom
  - These files will be located at .github/scripts
- Replaces the plugin for installing git hooks we had with a new one that supports reading from files
  - The new plugin also works on build, instead of needing to initialize, so the first build from the root directory will install the git hooks, simplifying the onboarding slightly
- Moves the .secrets file into .github/resources and renamed it to .sensitive-files to avoid naming conflicts with things that need a .secrets file (and better explain what it is for)
  - Adjusted the commit hook script to work with the new location/name

### What Should Reviewers Watch For?
<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->

If you're reviewing this PR, please check for these things in particular:
<!-- Add some items to the following list here -->
* Verify all PR security questions and checklists have been completed and addressed.


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [x] Yes
      * [ ] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @<!-- -->StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [x] Yes
      * [ ] No

### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A


### Submitter Checklist
<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

lsmitchell commented:
@StewGoin new plugin is:

https://github.com/rudikershaw/git-build-hook
https://mvnrepository.com/artifact/com.rudikershaw.gitbuildhook/git-build-hook-maven-plugin

Replacing:

https://github.com/phillipuniverse/githook-maven-plugin
https://mvnrepository.com/artifact/io.github.phillipuniverse/githook-maven-plugin

Didn't see any damning dependencies or anything in the new plugin. (Lots of its dependencies including slf4j are test scoped)
https://github.com/rudikershaw/git-build-hook/blob/develop/pom.xml

lsmitchell commented:
If accepted as-is, I'll update the wiki to note that to install the git hooks you just need to do a build from the apps directory instead of mvn initialize. With a fresh repo I pulled down, the .git folder is already added; there should be no reason to need initialize anymore with this change.

JeremyIR commented on .github/scripts/pre-commit.sh:
```diff
@@ -0,0 +1,34 @@
+
+  # for all files to be committed which are a secret file,
+  # grep for the header "$ANSIBLE_VAULT;1.1;AES256"
+  # if it's not there, abort.
```
This is really nitpicky, but the commend on lines 10-12 doesnt have a period at the end.

JeremyIR replied:
comment*

lsmitchell replied:
True, just copied it from the pom, but I'll fix this.

JeremyIR commented:
Looks good to me aside from consistency with comments ending in a period or not, if it matters that much

lsmitchell commented:
Just noticed the renamed .secrets file didn't make it into the commit, since it was being .gitignored; fixed the gitignore file and added the newly rehomed/rename secrets file.

JeremyIR commented:
@lsmitchell any advice on how to best test this? I know I have to clone your branch and install the plugin but any further guidance would be appreciated

lsmitchell commented:
If you grab this branch you can test it by doing the following:
1. Switch to the branch
2. Go to <bfd-repo>/.git/hooks and delete the pre-commit and pre-push hooks that are there
3. Maven build from the top level in apps (can skip tests)
4. Check that the hooks were installed into <bfd-repo>/.git/hooks

I did the following to test the pre-push hook is working:
once the hooks are installed go and change of the files listed in the sensitive-files (such as ops/terraform/services/base/values/test.eyaml) and remove the first line or replace everything with some plaintext, then try to commit. You should see it fail to commit due to the file being in plaintext. You can also see the pre-commit hook messages will work fine when committing the change.

JeremyIR commented:
Verifying on my end that the functionality works as properly. LGTM

StewGoin commented:
No concerns from me, proceed as planned.
